src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,191 @@
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        isElementIndexForAdd(index);
+        if (index == size) {
+            add(value);
+        } else {
+            linkBefore(value, node(index));
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        Object[] a = list.toArray();
+
+        for (Object value : a) {
+            final MyLinkedList.Node<T> l = tail;
+            final MyLinkedList.Node<T> newNode = new MyLinkedList.Node<T>(l, (T) value, null);
+            tail = newNode;
+            if (l == null) {
+                head = newNode;
+            } else {
+                l.next = newNode;
+            }
+            size++;
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        isElementIndex(index);
+        if (index < (size >> 1)) {
+            MyLinkedList.Node<T> x = head;
+            for (int i = 0; i < index; i++) {
+                x = x.next;
+            }
+            return x.value;
+        } else {
+            MyLinkedList.Node<T> x = tail;
+            for (int i = size - 1; i > index; i--) {
+                x = x.prev;
+            }
+            return x.value;
+        }
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        isElementIndex(index);
+        MyLinkedList.Node<T> x = node(index);
+        T oldVal = x.value;
+        x.value = value;
+        return oldVal;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        isElementIndex(index);
+        return unlink(node(index));
     }
 
     @Override
     public boolean remove(T object) {
+        if (object == null) {
+            for (MyLinkedList.Node<T> x = head; x != null; x = x.next) {
+                if (x.value == null) {
+                    unlink(x);
+                    return true;
+                }
+            }
+        } else {
+            for (MyLinkedList.Node<T> x = head; x != null; x = x.next) {
+                if (object.equals(x.value)) {
+                    unlink(x);
+                    return true;
+                }
+            }
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
+        if (size == 0) {
+            return true;
+        }
         return false;
     }
+
+    public void isElementIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Please enter right index from "
+                    + "0 to " + (size - 1));
+        }
+    }
+
+    public void isElementIndexForAdd(int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Please enter right index from "
+                    + "0 to " + (size - 1));
+        }
+    }
+
+    MyLinkedList.Node<T> node(int index) {
```

Comment:
```suggestion
    Node<T> getNode(int index) {
```
https://mate-academy.github.io/style-guides/java/java.html#s5.2.3-method-names